cloud/pkg/imageprepullcontroller/controller/downstream.go

```diff
@@ -0,0 +1,269 @@
+		if err != nil {
+			klog.Errorf("LabelSelector(%s) is not valid: %v", imagePrepull.Spec.ImagePrepullTemplate.LabelSelector, err)
+			return
+		}
+
+		nodes, err := dc.informer.Core().V1().Nodes().Lister().List(selector)
+		if err != nil {
+			klog.Errorf("Failed to get nodes with label %s: %v", selector.String(), err)
+			return
+		}
+
+		for _, node := range nodes {
+			if validateNode(node) {
+				nodesToPrepullImage = append(nodesToPrepullImage, node.Name)
+			}
+		}
+	}
+
+	// deduplicate: remove duplicate nodes to avoid repeating prepull images on the same node
+	nodesToPrepullImage = util.RemoveDuplicateElement(nodesToPrepullImage)
+
+	klog.Infof("Filtered finished, images will be prepulled on below nodes\n%v\n", nodesToPrepullImage)
+
+	go func() {
+		for _, node := range nodesToPrepullImage {
+			dc.processPrepull(node, imagePrepull)
+		}
+	}()
+}
+
+// imagePrepullJobDeleted is used to process deleted ImagePrepullJob in apiServer
+func (dc *DownstreamController) imagePrepullJobDeleted(imagePrepull *v1alpha1.ImagePrepullJob) {
+	// delete drom cache map
+	dc.imagePrepullJobManager.ImagePrepullMap.Delete(imagePrepull.Name)
+}
+
+// imagePrepullJobUpdate is used to process update of ImagePrepull in apiServer
+// Now we don't allow update spec, so we only update the cache map in imagePrepullJobUpdate func.
+func (dc *DownstreamController) imagePrepullJobUpdate(imagePrepull *v1alpha1.ImagePrepullJob) {
+	_, ok := dc.imagePrepullJobManager.ImagePrepullMap.Load(imagePrepull.Name)
+	// store in cache map
+	dc.imagePrepullJobManager.ImagePrepullMap.Store(imagePrepull.Name, imagePrepull)
+	if !ok {
+		klog.Infof("ImagePrepull Job %s not exist, and store it into first", imagePrepull.Name)
+		// If ImagePrepullJob not present in ImagePrepull map means it is not modified and added.
+		dc.imagePrepullJobAdded(imagePrepull)
+	}
+}
+
+// processPrepull process prepull job added and send it to edge nodes.
+func (dc *DownstreamController) processPrepull(node string, imagePrepull *v1alpha1.ImagePrepullJob) {
+	klog.V(4).Infof("begin to send imagePrepull message to %s", node)
+
+	imagePrepullTemplateInfo := imagePrepull.Spec.ImagePrepullTemplate
+	imagePrepullRequest := commontypes.ImagePrepullJobRequest{
+		Images:     imagePrepullTemplateInfo.Images,
+		NodeName:   node,
+		Secret:     imagePrepullTemplateInfo.ImageSecret,
+		RetryTimes: imagePrepullTemplateInfo.RetryTimesOnEachNode,
+		CheckItems: imagePrepullTemplateInfo.CheckItems,
+	}
+
+	// send prepull msg to edge node
+	msg := model.NewMessage("")
+	resource := buildPrepullResource(imagePrepull.Name, node)
+	msg.BuildRouter(modules.ImagePrepullControllerModuleName, modules.ImagePrepullControllerModuleGroup, resource, ImagePrepull).
+		FillBody(imagePrepullRequest)
+
+	err := dc.messageLayer.Send(*msg)
+	if err != nil {
+		klog.Errorf("Failed to send prepull message %s due to error %v", msg.GetID(), err)
+		return
+	}
```

Comment:
> Will the status be lost if only error messages are printed here? And at this time there is no any change in CR, users can only see the log of the CloudCore to troubleshoot problems.

I think we can put the timeout processing func before send msg to edge, so that if there is an error here, the timeout status will be updated later.

Reply:
done.